core/src/main/scala/com/metamx/tranquility/beam/ClusteredBeam.scala

```diff
@@ -327,16 +381,35 @@ class ClusteredBeam[EventType: Timestamper, InnerBeamType <: Beam[EventType]](
           beamMergeFn(
             (0 until tuning.partitions) map {
               partition =>
-                beamDecorateFn(bucket, partition)(new NoopBeam[EventType])
+                beamDecorateFn(beamPair._1, partition)(new NoopBeam[EventType])
             }
           )
         )
     }
   }
 
+  def groupEvents(event: EventType): (Interval, Option[Beam[EventType]]) = {
+    val eventTimestamp = timestamper(event)
+    if (beams.headOption.nonEmpty && beams.headOption.get.getInterval().get.contains(eventTimestamp)) {
```

Comment:
Why not do a loop over beams in reverse order?  The first one should get hit most of the time and then it just breaks out.